### PR TITLE
Initial implementation of IntrinsicArrayFunction

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**

### CODE REVIEW

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 1]
```diff
@@ -0,0 +1,1585 @@
...
+enum class IntrinsicArrayFunctions : int64_t {
+    Any,
+    MaxLoc,
+    MaxVal,
+    Merge,
+    MinLoc,
+    MinVal,
+    Product,
+    Shape,
+    Sum,
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_array_function_registry.h:28]: Should I add or revert any other function?
@certik

File: tests/reference/asr-intrinsics_27-11ba39d.stdout [hunk 2]
```diff
@@ -43,9 +43,30 @@
                     []
                     [(Print
                         ()
-                        [(IntrinsicFunction
+                        [(IntrinsicArrayFunction
                             Shape
-                            [(Var 2 a)]
+                            [(ArrayPhysicalCast
+                                (Var 2 a)
+                                FixedSizeArray
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/asr-intrinsics_27-11ba39d.stdout:50]: @czgdp1807
To which we have to apply `ArrayPhysicalCast`? Only constant array?
Is this correct change?

**Comment by Thirumalai-Shaktivel** [comment]
Ready for review!

**Review by certik — APPROVED** [review]
I think that this is good. 

@czgdp1807 can you please review this as well?

**Comment by certik** [comment]
After this is merged, we'll rename IntrinsicFunction to IntrinsicScalarFunction.

When IntrinsicScalarFunction is used on an array, we will also insert an explicit Broadcast node, to make it explicit that it's a scalar function being "vectorized".

File: src/libasr/pass/intrinsic_array_function_registry.h (hunk 1 above)
Comment (certik) [reply src/libasr/pass/intrinsic_array_function_registry.h:28]: I think this looks great.

**Review by czgdp1807 — APPROVED** [review]
Overall this is fine. However, I would like you to do the same changes in LPython (for making sure these changes work with LPython tests). It shouldn't take more than 2-3 hours of work (unless you encounter a blocker in LPython with these changes).

**Comment by certik** [comment]
@Thirumalai-Shaktivel if you could do this change in LPython as well, that would be great. Then we can merge both PRs.

**Comment by Thirumalai-Shaktivel** [comment]
Perfect! I will submit a PR with these changes in LPython.

Thanks for the review!

**Comment by Thirumalai-Shaktivel** [comment]
https://github.com/lcompilers/lpython/pull/2269

**Comment by certik** [comment]
Thanks! I merged both. Looks good.

**Pull request merged**